proxy/proxy/client.go

```diff
@@ -292,33 +241,129 @@ func isExpired(cfg *tls.Config) bool {
 	return time.Now().After(cfg.Certificates[0].Leaf.NotAfter)
 }
 
-func (c *Client) cachedCfg(instance string) (string, *tls.Config, string) {
+// startRefresh kicks off a refreshCfg asynchronously, returning a channel that is closed
+// when it's complete and updates the cacheEntry when complete. This function should only
+// be called from the scope of "cachedCfg", which controls the logic around throttling refreshes.
+func (c *Client) startRefresh(instance string, refreshCfgBuffer time.Duration) chan struct{} {
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		addr, cfg, ver, err := c.refreshCfg(instance)
+
+		c.cacheL.Lock()
+		old := c.cfgCache[instance]
+		// if we failed to refresh cfg do not throw out potentially valid one
+		if err != nil && !isExpired(old.cfg) {
+			logging.Errorf("failed to refresh the ephemeral certificate for %s, returning previous cert instead: %v", instance, err)
+			addr, cfg, ver, err = old.addr, old.cfg, old.version, old.err
```

Comment:
In this case, do we know `old.err` to be nil? If so, it would be better to assign `err` to `nil` directly.

Reply:
We don't know if `old.err` is `nil`. It doesn't really matter - if our current result failed, the old result is either good or bad. It's at least better than the current result.